### DISCUSSION TITLE
Non-localizable controls

### DISCUSSION THREAD

**Discussion by a-sanders** [perpetualKid/GetText.NET#24] [Ideas]
Hello,
It would be good to have the ability to mark some control as "non-localizable" for the Localizer (GetText.WindowsForms).

Regards.

**Comment by perpetualKid** [comment]
(admin note)
I've moved that to discussion section (which is new btw) to have further discussion, instead of having that an open issue.
(/admin note)

It would be great to understand the scenario or use case.

**Reply by a-sanders** [reply]
> It would be great to understand the scenario or use case.

This is my case.
There is the form - MainForm. There is a TextBox (read-only, disabled) on the form. It has empty Text property on start.

        private readonly ObjectPropertiesStore store = new ObjectPropertiesStore();

        public MainForm()
        {
            InitializeComponent();
            Localizer.Revert(this, store);
            LocalizeForm();
        }

        private void OnLocaleChanged(object sender, EventArgs e)
        {
            string locale = (cbLanguages.SelectedItem as Language).Locale;
            System.Threading.Thread.CurrentThread.CurrentUICulture = new CultureInfo(locale);
            Localizer.Revert(this, store);
            LocalizeForm();
        }

        private void LocalizeForm()
        {
            Localizer.Localize(this, catalog, store);
        }

By some logic during the program running, the textbox displays some text that shouldn't be localized.
But in a case when the user changes UI language the text box will get an empty value. More precisely, the value that was before the method call.

**Reply by perpetualKid** [reply]
for the time being you may just save and restore the particular textbox' text when you localize the form, such as 
```csharp
    private void LocalizeForm()
    {
        string textToStore = someTextBox.Text;
        Localizer.Localize(this, catalog, store);
        someTextBox.Text = string textToStore;
    }
```

as for a real solution, I currently could see two options:
- having a convention to use the Controls.Tag property and look for some specific value there, such as a particular string like "DoNotLocalize". This should with minimal effort for all controls. Downside is that you can't use the Tag property for anything else when needed for the DoNotLocalize.
- the other option could be to implement a specific DoNotLocalize attribute, which could be applied to to the control. Seems a more clean approach to me, but will require more effort to implement first. Also, when appyling, this would need to be applied in the Form.Designer.cs part, which typically is not touched since it's mostly auto-generated code. However the control variables definition is outside the auto-generated part, so should be impacted from Windows Forms Designer code generation.

Also, either approach would cover the whole control, so when using .Text and i.e. some ToolTip in addition, both properties would not be updated/localized.

**Reply by a-sanders** [reply]
>     * to use the Controls.Tag property
this is not a good idea and you mentioned this too

IMHO the simplest way at the moment - create a property IgnoreControls in the Localizer class that will store the controls' ref list.
So a developer could add (in form constructor) controls that should be skipped during form localizing.

**Reply by a-sanders** [reply]
.... BTW, the same way can be applied to specify a list of additional properties for localization, like EmptyListMsg...

**Reply by a-sanders** [reply]
I noticed that the problem is even wider than I thought. In fact, any TextBox controls will be revert to their original condition after 
Localizer.Revert(this, store)
have called. It looks like wrong logic at all. 
I guess any user-input controls should not be localized as default.
Does it make sense?

**Reply by perpetualKid** [reply]
> .... BTW, the same way can be applied to specify a list of additional properties for localization, like EmptyListMsg...

that's what I thougth as well, just providing some list of properties (names) to the Localizer. The challenge however is the extractor...

**Reply by perpetualKid** [reply]
> I noticed that the problem is even wider than I thought. In fact, any TextBox controls will be revert to their original condition after
> Localizer.Revert(this, store)
> have called. It looks like wrong logic at all.
> I guess any user-input controls should not be localized as default.
> Does it make sense?

I can clearly see where/why this is failing for you. 
The typical use case for the forms licalizer is to load localization once (at startup), not to change frequently at runtime.

In your scenario, it might be better to not use the localizer, but only assigned the required localizations manually
```csharp
label1.Text = catalog.GetString(label1Text);
//textBox1.Text // not translated
etc.
```
At the very end, the localizer is just encapsulating to iterate all controls and load the translation from catalog.GetString*.
What you are asking is some advanced customization which controls/properties are localized or not.

**Reply by a-sanders** [reply]
> The typical use case for the forms licalizer is to load localization once (at startup), not to change frequently at runtime.

Initially, we are not talking about changing the interface language frequently...

I see at least two possible approaches when the user changes the language by selecting the appropriate language with some control (radio button, combo box, menu, etc) - to require restarting the program or not. 
I prefer the second approach whenever possible.

The current logic of the localizer resets controls like text boxes and so I have to ask a user to restart a program or write some workarounds to prevent data loss...

Anyway, I sure the localizer should not touch text box controls because they use for user data input.

**Reply by perpetualKid** [reply]
"re-opening" ... seeing some more cases where this impacts usability, so will need to think about some improvements. Converting back to issue for now
